api/envoy/extensions/filters/udp/dns_filter/v3alpha/dns_filter.proto

```diff
@@ -44,15 +45,27 @@ message DnsFilterConfig {
   // in a client context. This message will contain the timeouts, retry,
   // and forwarding configuration for Envoy to make DNS requests to other
   // resolvers
+  //
+  // [#next-free-field: 6]
   message ClientContextConfig {
     // Sets the maximum time we will wait for the upstream query to complete
     // We allow 5s for the upstream resolution to complete, so the minimum
     // value here is 1. Note that the total latency for a failed query is the
     // number of retries multiplied by the resolver_timeout.
     google.protobuf.Duration resolver_timeout = 1 [(validate.rules).duration = {gte {seconds: 1}}];
 
+    // This field was used for the `upstream_resolvers` field in Envoy 1.18 and
+    // for `dns_resolution_config` in Envoy 1.19.
+    //
+    // Control planes should set this field for Envoy 1.19.0 and 1.19.1
+    // clients, and set `dns_resolution_config` instead for later Envoy clients.
+    // Control planes that need to simultaneously support Envoy 1.18 should
+    // avoid Envoy 1.19.0 and 1.19.1.
+    config.core.v3.DnsResolutionConfig compat_dns_resolution_config = 2
```

Comment:
Could you please provide a little bit more details about how this new field will be used?

From this code, only dns_resolution_config field is used. https://github.com/envoyproxy/envoy/blob/db74e313b3651588e59c671af45077714ac32cef/source/extensions/filters/udp/dns_filter/dns_filter.cc#L162

To my understanding, dns_resolution_config is used by cares DNS resolver to resolve targets.  Do you mean adding a new field(with identical type) here for same purpose, but for different version?

Reply:
referring this comment: https://github.com/envoyproxy/envoy/issues/17921#issuecomment-912050336
> we need a proto definition where field 2 isn't used for 2 different purposes.

So shouldn't this field actually be of type Address? Correct me if I am wrong here.

`    repeated config.core.v3.Address upstream_resolvers = 2;` [code link](https://github.com/envoyproxy/envoy/blob/release/v1.18/api/envoy/extensions/filters/udp/dns_filter/v3alpha/dns_filter.proto#L57)

Reply:
For me, migrating from 1.18, Address would be ideal. However 1.19 already shipped with the DNS config in field 2, so applications that support 1.19.0 will also need to migrate. It's a matter of who takes the burden of having to fork the protobufs to get the type they need in the right field.

If we have consensus that field 2 should be of type Address, that's great for me because it's a lot less work :)

Reply:
But you make a good point. I ought to err on making it easier for myself, at least until Envoy API reviewers give their feedback :)

Updated the PR  with Address.

Reply:
Just thinking about the other alternates, can we have v3alpha and v3 support simultaneously? Extrapolating from this [comment](https://github.com/envoyproxy/envoy/issues/17921#issuecomment-909713933) from Matt, assuming this is what was mean from the comment.

The Control plane would support both the versions v3alpha & v3 for DnsFilter. DnsFilter v3alpha support for v1.18 envoy (and prior to v1.18) & DnsFilter v3 support for v1.19.2 and future releases (v1.20+).

This way the envoys currently using DnsFilter can migrate from 1.18 to strictly 1.19.2+

Reply:
I don't need a single Envoy release to simultaneously support multiple versions of the dns filter. I just need the protobufs to be defined in such a way that I can serve both 1.18 and 1.19 Envoy clients.

So rather than changing v3alpha definitions, Envoy 1.19 could have frozen the v3alpha protobufs and moved the implementation to v3alpha1. Then I could have served v3alpha to Envoy 1.18 and v3alpha1 to Envoy 1.19. But right now, this PR is just trying to find the least worst way to recover.